### PR TITLE
[Named min timestamp leases] Metrics

### PR DESCRIPTION
**Pull request by ergo14**
Metrics

### CODE REVIEW

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java [hunk 1]
```diff
@@ -158,22 +174,28 @@ private boolean wasRequestFullyFulfilled(RequestAndResponse requestAndResponse)
...
+                .filter(timestampName -> {
+                    int requestedTimestamps = requests.get(timestampName);
+                    long returnedTimestamps = responses.get(timestampName).getFreshTimestamps().getCount();
+                    if (returnedTimestamps >= requestedTimestamps) {
+                        return false;
+                    }
+                    telemetryTimestampShortFallConsumer.accept(requestedTimestamps - returnedTimestamps);
+                    return true;
+                })
+                .count();
```
Comment (ergo14) [inline lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: This code path will be moved and factored as part of the previous PR reviews, so just review the logic here.

I'll move things later as appropriate

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java [hunk 2]
```diff
@@ -158,22 +174,28 @@ private boolean wasRequestFullyFulfilled(RequestAndResponse requestAndResponse)
...
-            return returnedTimestamps >= requestedTimestamps;
-        });
+        long wereNotFullyFulfilled = requests.keySet().stream()
+                .filter(timestampName -> {
+                    int requestedTimestamps = requests.get(timestampName);
+                    long returnedTimestamps = responses.get(timestampName).getFreshTimestamps().getCount();
+                    if (returnedTimestamps >= requestedTimestamps) {
+                        return false;
+                    }
+                    telemetryTimestampShortFallConsumer.accept(requestedTimestamps - returnedTimestamps);
```
Comment (ergo14) [inline lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: I only want to report shortfall > 0

An alternative is to report shortfall >= 0, or any possible shortfall (or overflow)

Comment (ergo14) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: Realistically, we care about the count here rather than the histogram. But, it may be needed eventually

File: lock-api/src/main/metrics/metric-schema.yml [hunk 3]
```diff
@@ -30,4 +30,10 @@ namespaces:
         type: timer
         docs: observed call duration during leader election
 
+  timestampLease:
```
Comment (ergo14) [inline lock-api/src/main/metrics/metric-schema.yml:33]: This will generate a class named the same as the other metrics in this PR. I think that's fine

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java [hunk 4]
```diff
@@ -24,16 +25,33 @@
 import java.util.SortedMap;
 import java.util.TreeMap;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
 import javax.annotation.concurrent.GuardedBy;
 
 final class NamedMinTimestampTrackerImpl implements NamedMinTimestampTracker {
     private final String timestampName;
+    private final AtomicInteger numLocksHeld;
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:34]: - Write path: this is modified in `lock` and `unlock` calls which are already synchronized
- Read path: this will be ran in some telemetry wc thread, and the read seems to be a volatile read only - I can't imagine a less perf-impacting read (other than a final field) if we are to have memory visibility. I guess we could have a variable with no synchronization and hope for the best, the metrics don't need to be 100% sure

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java [hunk 5]
```diff
@@ -24,16 +25,33 @@
...
+    private final AtomicLong approximateMinLocked;
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:35]: This one I only update when there are calls for `getMinimumTimestamp` as calling getFirstKey is O(tree height) in the tree map iirc

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java [hunk 6]
```diff
@@ -24,16 +25,33 @@
...
 
     @GuardedBy("this")
     private final SortedMap<Long, UUID> holdersByTimestamp = new TreeMap<>();
 
-    NamedMinTimestampTrackerImpl(String timestampName) {
+    private NamedMinTimestampTrackerImpl(
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:40]: I also considered adding meters for lock/unlock as generally those give more signal/granularity than a size metric

But, looking at `Meter`, updating it involves more CASs etc

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java [hunk 7]
```diff
@@ -46,6 +64,7 @@ public synchronized void lock(long timestamp, UUID requestId) {
                     SafeArg.of("requestId", requestId),
```
Comment (ergo14) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:64]: One thing to note is: the code path is now different than before as this impacts the immutable timestamp. I can special case.

Of course, with all concurrency comments on this PR, I may be too paranoid here

File: timelock-impl/src/main/metrics/timestampleases.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+namespaces:
+  timestampLease:
+    docs: Metrics tracking TimeLock's implementation of timestamp leases
+    metrics:
+      locksHeld:
+        docs: Number of locks held in timestamp trackers
+        type: gauge
+        tags:
+          - name
+      approximateSmallestLeased:
```
Comment (ergo14) [inline timelock-impl/src/main/metrics/timestampleases.yml:13]: name to make sure any operator/dev knows this is best-effort and does not over-index

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java (hunk 2 above)
Comment (jkozlowski) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: Why do we need a histogram?

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java [hunk 9]
```diff
@@ -61,24 +63,38 @@ public final class TimestampLeaseAcquirerImpl implements TimestampLeaseAcquirer
     private final NamespacedTimestampLeaseService delegate;
     private final Unlocker unlocker;
     private final Supplier<UUID> uuidSupplier;
+    private final Consumer<Long> telemetryTimestampShortFallConsumer;
 
     private final Retryer<RequestAndResponse> retryer = RetryerBuilder.<RequestAndResponse>newBuilder()
             .retryIfResult(requestAndResponse -> !wasRequestFullyFulfilled(requestAndResponse))
             .retryIfException()
             .withStopStrategy(StopStrategies.stopAfterAttempt(3))
```
Comment (jkozlowski) [inline lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: Should we add some backoff?

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java [hunk 10]
```diff
@@ -158,22 +174,28 @@ private boolean wasRequestFullyFulfilled(RequestAndResponse requestAndResponse)
...
 
-        boolean wasFullyFulfilled = requests.keySet().stream().allMatch(timestampName -> {
-            int requestedTimestamps = requests.get(timestampName);
-            long returnedTimestamps =
-                    responses.get(timestampName).getFreshTimestamps().getCount();
-            return returnedTimestamps >= requestedTimestamps;
-        });
+        long wereNotFullyFulfilled = requests.keySet().stream()
+                .filter(timestampName -> {
+                    int requestedTimestamps = requests.get(timestampName);
```
Comment (jkozlowski) [inline lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: Do we validate somewhere that these are non-zero?

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java (hunk 2 above)
Comment (jkozlowski) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: I think we will struggle with interpreting these. If you care about magnitude, just log it. But I'd just log a binary based on ```wasFullyFulfilled > 0```

Comment (jkozlowski) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: and if that turns out to be false and we need more, we'll do that later, but at this point I'm struggling to think why I'd want to see the exact number.

Comment (ergo14) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: > Why do we need a histogram?

More data? Most appropriate would be a meter then? I don't usually like counters because of how they are reported but can use that as well

Comment (jkozlowski) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: yeah, just a counter

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java (hunk 9 above)
Comment (ergo14) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: Discussed offline that we won't be adding this now

File: lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java (hunk 10 above)
Comment (ergo14) [reply lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java]: No I can add a preconditions before I start the whole retry block?

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java [hunk 11]
```diff
@@ -58,14 +77,17 @@ public synchronized void unlock(long timestamp, UUID requestId) {
...
+        numLocksHeld.decrementAndGet();
     }
 
     @Override
     public synchronized Optional<Long> getMinimumTimestamp() {
         if (holdersByTimestamp.isEmpty()) {
             return Optional.empty();
         }
-        return Optional.of(holdersByTimestamp.firstKey());
+        Long minimum = holdersByTimestamp.firstKey();
```
Comment (jkozlowski) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:88]: Can we just report it in the same way we do right now in the immutable timestamp case? Just have it be done by Atlas and in the consumer application. Doing this in timelock breaks the usecase here of being able to write dashboard for a particular consumer (or at least maybe makes it more snowflakey/different than how we do all other metrics?)

Comment (jkozlowski) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:88]: And in general, I'm less bought in to the idea of this metric right now / care about it the least for the initial milestone, so at least split this out into a separate PR, so we can ship the counter first.

Comment (ergo14) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:88]: I will spin out a PR **only** for the shortfall thing

Comment (ergo14) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/NamedMinTimestampTrackerImpl.java:88]: https://github.com/palantir/atlasdb/pull/7402

**Comment by ergo14** [comment]
Setting to draft - need to:
- Spin to 2 PRs
- One for the size of the lock collection on TimeLock
- One where I move the metric for the value of the named timestamp to the client side
